README.md
# node-red-contrib-ui-popup Readme

node-red-contrib-ui-popup is a collection of popup-dialogues for the Node-RED
Dashboard

## Install

### Directly in Node-RED via palette manager

Start Node-RED

Select 'Manage palette' from menu top-right

Click on the Install tab

Search for ui-popup in the searchbar and find node-red-contrib-ui-popup

### Manual install

Run command on Node-RED installation directory.

    npm install @icke1983/node-red-contrib-ui-popup

## Usage

This is a collection of popup-dialogues for the node-red dashboard.

The following dialogues are available.

### timepicker node

![timepicker Example](https://raw.githubusercontent.com/icke1983/node-red-contrib-ui-popup/master/images/timepicker.PNG)

### colorpicker node

![colorpicker Example](https://raw.githubusercontent.com/icke1983/node-red-contrib-ui-popup/master/images/colorpicker.PNG)

### colortemp node

![colortemp Example](https://raw.githubusercontent.com/icke1983/node-red-contrib-ui-popup/master/images/colortemp.PNG)

### simpleslider node

![simpleslider Example](https://raw.githubusercontent.com/icke1983/node-red-contrib-ui-popup/master/images/simpleslider.PNG)

### dropdown node

![dropdown Example](https://raw.githubusercontent.com/icke1983/node-red-contrib-ui-popup/master/images/dropdown.PNG)

### iframe node

![iframe Example](https://raw.githubusercontent.com/icke1983/node-red-contrib-ui-popup/master/images/iframe.PNG)

## Inspiration and thanking

- the dropdown node is inspired by [node-red-contrib-ui-contextmenu](https://github.com/bartbutenaers/node-red-contrib-ui-contextmenu) - thank to [Bart Butenaers](https://github.com/bartbutenaers)

- the colortemp node contains calculating code of [Tanner Helland](https://tannerhelland.com/2012/09/18/convert-temperature-rgb-algorithm-code.html) and [Mike D Sutton](https://gist.github.com/EDais) - thank for sharing of your knowledge

## License

The MIT License


CHANGELOG.md
# node-red-contrib-ui-popup Changelog

## 0.9.9

### fix:
- timepicker node: storing typeOfTopic
- colorpicker node: storing typeOfTopic
- colortemp node: storing typeOfTopic
- simpleslider node: storing typeOfTopic
- dropdown node: storing typeOfTopic

### improvements:
- add: iframe node

## 0.9.0

- first official pblishing

ui_popup_dropdown.js
module.exports = function(RED) {
    var ui = undefined;
    
    var generateHTML = function (node, config) {
        
        var id = node.id.replace(/[^\w]/g, "");
        var pos = "";
        var msgPosX = "";
        var msgPosY = "";
        var li = "";
        config.options.forEach(function(item, index) {
            li += String.raw`<li ng-click="sendPayload(${index})"><span class="ui-popup-dropdown-icon">`;
            if (item.icon != "") { li += String.raw`<ui-icon icon="${item.icon}"></ui-icon>`; }
            li += String.raw`</span><span class="ui-popup-dropdown-label">${item.label}</span></li>`;
        });
        var HTML = "";
        switch(config.position) {
            case "tl":
                pos = "top: 0px; left: 0px; margin-top: 0px; margin-left: 0px;";
                break;
            case "tr":
                pos = "top: 0px; right: 0px; margin-top: 0px; margin-right: 0px;";
                break;
            case "bl":
                pos = "bottom: 0px; left: 0px; margin-bottom: 0px; margin-left: 0px;";
                break;
            case "br":
                pos = "bottom: 0px; right: 0px; margin-bottom: 0px; margin-right: 0px;";
                break;
            case "center":
                pos = "top: 50%; left: 50%; transform: translate(-50%, -50%); margin: 0px;";
                break;
            case "man":
                let posX;
                let posY;
                if (config.posXType == "num") {
                    posX = config.posX;
                } else{
                    posX = "0";
                    msgPosX = String.raw` msgposx="${config.posX}"`;
                }
                if (config.posYType == "num") {
                    posY = config.posY;
                } else {
                    posY = "0";
                    msgPosY = String.raw` msgposy="${config.posY}"`;
                }
                pos = "top: "+posY+"px; left: "+posX+"px; margin-top: 0px; margin-left: 0px;";
                break;
        } 
        HTML += String.raw`
            <style id="ui-popup-dropdown-style_${id}">
                .ui-popup-dialog._${id} {
                    ${pos}
                }
            </style>
            <style id="ui-popup-dialog-style">
                .ui-popup-dialog {
                    padding: 0px;
                    position: relative;
                }
                .ui-popup-dialog[open] {
                    background: none;
                    border: none;
                }
                
                .ui-popup-dialog[open]::backdrop {
                    background: rgba(0,0,0,0.6);
                    backdrop-filter: blur(3px);
                }
            </style>
            <style id="ui-popup-dropdown-style">
                .ui-popup-dropdown {
                    position: absolute;
                }
                .ui-popup-dropdown .content{
                    width: auto;
                }
                .ui-popup-dropdown .content *{
                    box-sizing: border-box;
                }
                .ui-popup-dropdown ul {
                    list-style-type: none;
                    margin: 0px;
                    padding: 0px;
                }
                .ui-popup-dropdown li{
                    display: flex;
                    padding: 5px;
                    color: var(--nr-dashboard-widgetTextColor);
                }
                .ui-popup-dropdown li:hover{
                    background-color: var(--nr-dashboard-widgetColor);
                    cursor: pointer;
                }
                .ui-popup-dropdown .ui-popup-dropdown-icon {
                    display: inline-block;
                    text-align: center;
                    margin-right: 5px;
                    padding-right: 5px;
                    width: 1.5em;
                }
            </style>
            <div id="ui-popup-dropdown_${id}" class="nr-dashboard-cardpanel ${config.class}"${msgPosX}${msgPosY}>
                <div class="content">
                    <ul>
                        ${li}
                    </ul>
                </div>
            </div>`;
        return HTML;
    }
    
    function ui_popup_dropdown(config) {
        try {
            RED.nodes.createNode(this,config);
            var node = this;
            
            if (ui === undefined) {
                ui = RED.require("node-red-dashboard")(RED);
            }
            
            node.on('input', function(msg) {
                node.msg = msg;
            });

            var done = ui.addWidget({
                node: node,
                format: generateHTML(node, config),
                templateScope: "local",
                group: config.group,
                emitOnlyNewValues: false,
                forwardInputMessages: false,
                storeFrontEndInputAsState: false,
                group: config.group,
                order: config.order,
                width: 0,
                height: -1,
                
                convertBack: function (value) {
                    return value;
                },

                beforeSend: function (msg, orig) {
                    if(orig) {
                        let index = orig.msg.ui_popup_dropdown_item
                        let newPayload;
                        //manipulate topic
                        switch (config.options[index].topicType) {
                            case "msg":
                                orig.msg.topic = ""
                                try { orig.msg.topic = RED.util.getMessageProperty(node.msg, config.options[index].topic); } catch (e) {}
                                break;
                            case "str":
                                orig.msg.topic = config.options[index].topic;
                                break;
                        }
                        switch (config.options[index].payloadType) {
                            case "msg":
                                newPayload = "";
                                try { newPayload = RED.util.getMessageProperty(node.msg, config.options[index].payload); } catch (e) {}
                                break;
                            case "str":
                                newPayload = config.options[index].payload;
                                break;
                            case "num":
                                newPayload = parseFloat(config.options[index].payload);
                                break;
                            case "bool":
                                newPayload = Boolean(config.options[index].payload);
                                break;
                        }
                        RED.util.setMessageProperty(orig.msg, config.sendValueTo, newPayload, true);
                        
                        delete orig.msg.ui_popup_dropdown_item;
                        return orig.msg;
                    }     
                },

                initController: function ($scope, events) {
                    let olddialog = document.querySelectorAll("#ui-popup-dialog_"+id);
                    olddialog.forEach(function(dialog) { dialog.remove(); } );
                    $scope.dropdown = document.createElement("dialog");
                    $scope.msgPosX = null;
                    $scope.msgPosY = null;
                    setTimeout(function () {
                        //close if click outside
                        let onClick = function(event) {
                            if (event.target === $scope.dropdown) {
                                $scope.dropdown.close();
                            }
                        }
                        let dialog_style = document.querySelector("#ui-popup-dialog-style");
                        let dropdown_style = document.querySelector("#ui-popup-dropdown-style");
                        let dropdown_style_id = document.querySelector("#ui-popup-dropdown-style_"+id);
                        let cardpanel = document.querySelector("#ui-popup-dropdown_" + id);
                        let cardPanelContainer = document.createElement("ui-card-panel");
                        $scope.msgPosX = cardpanel.getAttribute("msgposx");
                        $scope.msgPosY = cardpanel.getAttribute("msgposy");
                        document.head.appendChild(dialog_style);
                        document.head.appendChild(dropdown_style);
                        document.head.appendChild(dropdown_style_id);
                        cardpanel.parentElement.remove();
                        cardPanelContainer.appendChild(cardpanel);
                        cardPanelContainer.classList.add("ui-popup-dropdown");
                        $scope.dropdown.setAttribute("id", "ui-popup-dialog_"+id);
                        $scope.dropdown.appendChild(cardPanelContainer);
                        $scope.dropdown.classList.add("ui-popup-dialog");
                        $scope.dropdown.classList.add("_"+id);
                        $scope.dropdown.addEventListener("click", onClick);
                        document.body.appendChild($scope.dropdown);
                        
                    }, 300);
                    
                    $scope.sendPayload = function (value) {
                        $scope.dropdown.close();
                        $scope.send({ui_popup_dropdown_item: value})
                    }

                    $scope.$watch("msg", function (msg) {
                        if (!msg) { 
                            return;
                        } else {
                            
                            $scope.dropdown.close();
                            $scope.dropdown.showModal();
                            $scope.dropdown.style.height = $scope.dropdown.firstElementChild.offsetHeight.toString()+"px";
                            $scope.dropdown.style.width = $scope.dropdown.firstElementChild.offsetWidth.toString()+"px";
                            if ($scope.msgPosX != null) {
                                if(parseInt(msg.ui_popup_dropdown_posX) + $scope.dropdown.offsetWidth >= window.innerWidth) {
                                    $scope.dropdown.style.left = (window.innerWidth - $scope.dropdown.offsetWidth).toString()+"px";
                                } else {
                                    $scope.dropdown.style.left = msg.ui_popup_dropdown_posX+"px";
                                }
                            }
                            if ($scope.msgPosY != null) {
                                if(parseInt(msg.ui_popup_dropdown_posy) + $scope.dropdown.offsetHeight >= window.innerHeight) {
                                    $scope.dropdown.style.top = (window.innerHeight - $scope.dropdown.offsetHeight).toString()+"px";
                                } else {
                                    $scope.dropdown.style.top = msg.ui_popup_dropdown_posY+"px";
                                }
                            }
                            $scope.payload = msg.payload;
                            
                        };                        
                    });
                },
                
                beforeEmit: function (msg, value) {
                    var newMsg = {};
		            if (msg) {
                        newMsg = msg;
                        newMsg.ui_popup_dropdown_posX = "0";
                        newMsg.ui_popup_dropdown_posY = "0"
                        try { newMsg.ui_popup_dropdown_posX = RED.util.getMessageProperty(msg, config.posX).toString(); } catch (e) {}
                        try { newMsg.ui_popup_dropdown_posY = RED.util.getMessageProperty(msg, config.posY).toString(); } catch (e) {}
                    }
                    return { msg: newMsg };
                
            }

            });
        }
        catch (e) {
            //console.log(e);
        }
        node.on("close", done);
    }
    RED.nodes.registerType("ui_popup-dropdown",ui_popup_dropdown);
}

ui_popup_simpleslider.js
module.exports = function(RED) {
    var ui = undefined;
    
    var generateHTML = function (node, config) {
        var id = node.id.replace(/[^\w]/g, "");
        var pos = "";
        var HTML = "";
        switch(config.position) {
            case "tl":
                pos = "top: 0px; left: 0px; margin-top: 0px; margin-left: 0px;";
                break;
            case "tr":
                pos = "top: 0px; right: 0px; margin-top: 0px; margin-right: 0px;";
                break;
            case "bl":
                pos = "bottom: 0px; left: 0px; margin-bottom: 0px; margin-left: 0px;";
                break;
            case "br":
                pos = "bottom: 0px; right: 0px; margin-bottom: 0px; margin-right: 0px;";
                break;
            case "center":
                pos = "top: 50%; left: 50%; transform: translate(-50%, -50%); margin: 0px;";
                break;
        } 
        HTML += String.raw`
            <style id="ui-popup-simpleslider-style_${id}">
                .ui-popup-dialog._${id} {
                    ${pos}
                }
            </style>
            <style id="ui-popup-dialog-style">
                .ui-popup-dialog {
                    padding: 0px;
                    position: relative;
                }
                .ui-popup-dialog[open] {
                    background: none;
                    border: none;
                }
                
                .ui-popup-dialog[open]::backdrop {
                    background: rgba(0,0,0,0.6);
                    backdrop-filter: blur(3px);
                }
            </style>
            <style id="ui-popup-simpleslider-style">
                .ui-popup-simpleslider {
                    position: absolute;//z-index: 99999;
                }
                .ui-popup-simpleslider .content{
                    width: 650px;
                    display: grid;
                    grid-template-columns: 75px auto 75px;
                }
                .ui-popup-simpleslider span{
                    color: var(--nr-dashboard-widgetTextColor);
                }
                .ui-popup-simpleslider .ui-popup-slider {
                    -webkit-appearance: none;
                    width: 100%;
                    height: 25px;
                    border-radius: 12px;
                    outline: none;
                    -webkit-transition: .2s;
                    transition: opacity .2s;
                    background: -moz-linear-gradient(right, ${config.colorStart} 0%, ${config.colorEnd} 100%);
                    background: -ms-linear-gradient(right, ${config.colorStart} 0%, ${config.colorEnd} 100%);
                    background: -o-linear-gradient(right, ${config.colorStart} 0%, ${config.colorEnd} 100%);
                    background: -webkit-linear-gradient(left, ${config.colorStart} 0%, ${config.colorEnd} 100%);
                    background: linear-gradient(to right, ${config.colorStart} 0%, ${config.colorEnd} 100%);
                }
                .ui-popup-simpleslider .ui-popup-slider::-webkit-slider-thumb {
                    -webkit-appearance: none;
                    appearance: none;
                    width: 25px;
                    height: 25px;
                    border-radius: 50%;
                    border: solid black 2px;
                    background: none;
                    cursor: pointer;
                }
                .ui-popup-simpleslider .ui-popup-slider::-moz-range-thumb {
                    width: 25px;
                    height: 25px;
                    border-radius: 50%;
                    border: solid black 2px;
                    background: none;
                    cursor: pointer;
                }
                .ui-popup-simpleslider .nr-dashboard-button {
                    margin: 6px;
                }
                .ui-popup-simpleslider .md-button{
                    font-size: 1em;
                }
            </style>
            <div id="ui-popup-simpleslider_${id}" class="nr-dashboard-cardpanel ${config.class}">
                <p class="nr-dashboard-cardtitle">${config.label}</p>
                <div class="content">
                    <div style="text-align: center;"><span>${config.sndLabel}</span></div>
                    <div>
                        <input type="range" min="${config.rangeStart}" max="${config.rangeEnd}" step="${config.rangeStep}" class="ui-popup-slider" ng-value="val" ng-model="val">
                    </div>
                    <div style="text-align: center;"><span ng-bind-html="val+'${config.unit}'"></span></div>
                 </div>
                <div style="display: grid; grid-template-columns: 1fr 1fr 1fr;">
                    <div class="nr-dashboard-button">
                        <span></span>
                    </div>
                    <div class="nr-dashboard-button">
                        <button class="md-button md-raised" ng-click="sliderCancel()"><span>${config.labelCancelButton}</span></button>
                    </div>
                    <div class="nr-dashboard-button">
                        <button class="md-button md-raised" ng-click="sliderOk()"><span>${config.labelOkButton}</span></button>
                    </div>
                </div>
            </div>`;
        return HTML;
    }
    
    function ui_popup_simpleslider(config) {
        try {
            RED.nodes.createNode(this,config);
            var node = this;
            
            if (ui === undefined) {
                ui = RED.require("node-red-dashboard")(RED);
            }
            
            node.on('input', function(msg) {
                                
                //set topic
                if (config.topicType == "msg") {
                    node.topic = "";
                    try { node.topic = RED.util.getMessageProperty(msg, config.topic); } catch (e) {}
                }
                //node.send({payload: msg.payload});
            });

            var done = ui.addWidget({
                node: node,
                format: generateHTML(node, config),
                templateScope: "local",
                group: config.group,
                emitOnlyNewValues: false,
                forwardInputMessages: false,
                storeFrontEndInputAsState: false,
                group: config.group,
                order: config.order,
                width: 0,
                height: -1,
                
                convertBack: function (value) {
                    return value;
                },

                beforeSend: function (msg, orig) {
                    if(orig) {
                        //manipulate topic
                        switch (config.topicType) {
                            case "msg":
                                orig.msg.topic = node.topic;
                                break;
                            case "str":
                                orig.msg.topic = config.topic;
                                break;
                        }
                        
                        try { RED.util.setMessageProperty(orig.msg, config.sendValueTo, orig.msg.ui_popup_val, true); } catch (e) {}
                        delete orig.msg.ui_popup_val;
                        return orig.msg;
                    }     
                },

                initController: function ($scope, events) {
                    let olddialog = document.querySelectorAll("#ui-popup-dialog_"+id);
                    olddialog.forEach(function(dialog) { dialog.remove(); } );
                    $scope.simpleslider = document.createElement("dialog");                   
                    $scope.val = 0;
                    setTimeout(function () {
                        //close if click outside
                        let onClick = function(event) {
                            if (event.target === $scope.simpleslider) {
                                $scope.simpleslider.close();
                            }
                        }
                        let dialog_style = document.querySelector("#ui-popup-dialog-style");
                        let simpleslider_style = document.querySelector("#ui-popup-simpleslider-style");
                        let simpleslider_style_id = document.querySelector("#ui-popup-simpleslider-style_"+id);
                        let cardpanel = document.querySelector("#ui-popup-simpleslider_" + id);
                        let cardPanelContainer = document.createElement("ui-card-panel");
                        document.head.appendChild(dialog_style);
                        document.head.appendChild(simpleslider_style);
                        document.head.appendChild(simpleslider_style_id);
                        cardpanel.parentElement.remove();
                        cardPanelContainer.appendChild(cardpanel);
                        cardPanelContainer.classList.add("ui-popup-simpleslider");
                        $scope.simpleslider.appendChild(cardPanelContainer);
                        $scope.simpleslider.setAttribute("id", "ui-popup-dialog_"+id);
                        $scope.simpleslider.classList.add("ui-popup-dialog");
                        $scope.simpleslider.classList.add("_"+id);
                        $scope.simpleslider.addEventListener("click", onClick);
                        document.body.appendChild($scope.simpleslider);
                    }, 300);
                    
                    $scope.sliderCancel = function () {
                        $scope.simpleslider.close();
                    }

                    $scope.sliderOk = function () {
                        $scope.sliderCancel();
                        $scope.send({ui_popup_val: $scope.val});
                    }
                   
                    $scope.$watch("msg", function (msg) {
                        if (!msg) { 
                            return;
                        } else {
                            if (msg.val != undefined) {
                                $scope.simpleslider.close();
                                $scope.simpleslider.showModal();
                                $scope.simpleslider.style.height = $scope.simpleslider.firstElementChild.offsetHeight.toString()+"px";
                                $scope.simpleslider.style.width = $scope.simpleslider.firstElementChild.offsetWidth.toString()+"px";
                                $scope.val = msg.val;
                            }
                        };                        
                    });
                },
                
                beforeEmit: function (msg, value) {
                    var newMsg = {};
		            if (msg) {
                        newMsg.socketid = msg.socketid;
                        newMsg.val = "0";
                        try { newMsg.val = RED.util.getMessageProperty(msg, config.payload); } catch (e) {}
                    }
                    return { msg: newMsg };
                
            }

            });
        }
        catch (e) {
            //console.log(e);
        }
        node.on("close", done);
    }
    RED.nodes.registerType("ui_popup-simpleslider",ui_popup_simpleslider);
}

ui_popup_iframe.js
module.exports = function(RED) {
    var ui = undefined;
    
    var generateHTML = function (node, config) {
        var id = node.id.replace(/[^\w]/g, "");
        var pos = "";var HTML = "";
        switch(config.position) {
            case "tl":
                pos = "top: 0px; left: 0px; margin-top: 0px; margin-left: 0px;";
                break;
            case "tr":
                pos = "top: 0px; right: 0px; margin-top: 0px; margin-right: 0px;";
                break;
            case "bl":
                pos = "bottom: 0px; left: 0px; margin-bottom: 0px; margin-left: 0px;";
                break;
            case "br":
                pos = "bottom: 0px; right: 0px; margin-bottom: 0px; margin-right: 0px;";
                break;
            case "center":
                pos = "top: 50%; left: 50%; transform: translate(-50%, -50%); margin: 0px;";
                break;
        } 
        HTML += String.raw`
            <style id="ui-popup-iframe-style_${id}">
                .ui-popup-dialog._${id} {
                    ${pos}
                    width: ${config.documentWidth};
                    height: ${config.documentHeight};
                    overflow: hidden
                }
            </style>
            <style id="ui-popup-dialog-style">
                .ui-popup-dialog {
                    padding: 0px;
                    position: relative;
                }
                .ui-popup-dialog[open] {
                    background: none;
                    border: none;
                }
                
                .ui-popup-dialog[open]::backdrop {
                    background: rgba(0,0,0,0.6);
                    backdrop-filter: blur(3px);
                }
            </style>
            <style id="ui-popup-iframe-style">
                .ui-popup-iframe {
                    position: absolute;
                    width: calc(100% - 2px);
                    height: calc(100% - 2px);
                }
                .ui-popup-iframe .nr-dashboard-cardpanel {
                    width: 100%;
                    height: 100%;
                    display: flex;
                    flex-flow: column;
                }

                .ui-popup-iframe .nr-dashboard-cardtitle {
                    flex: 0 1 auto;
                }

                .ui-popup-iframe .content{
                    flex: 1 1 auto;
                }
                .ui-popup-iframe iframe{
                    width: 100%;
                    height: 100%;
                    border: none;
                }
                .ui-popup-iframe .closeButton{
                    position: absolute;
                    right: 12px;
                    top: 10px;
                }
                .ui-popup-iframe .closeButton span{
                    font-size: x-large;
                }
                .ui-popup-iframe span{
                    color: var(--nr-dashboard-widgetTextColor);
                }
            </style>
            <div id="ui-popup-iframe_${id}" class="nr-dashboard-cardpanel ${config.class}">
                <p class="nr-dashboard-cardtitle">${config.label}</p>
                <div class="content" framesrc="${config.url}">
                    <iframe loading="lazy"><p>Ihr Browser kann leider keine eingebetteten Frames anzeigen. Sie können die eingebettete Seite <a href="${config.url}">hier</a> aufrufen!</iframe>
                    </p>
                </div>
                <div class="closeButton" ng-click="dialogClose()"><span><i class="fa fa-window-close-o"></i></span></div>
            </div>`;
        return HTML;
    }
    
    function ui_popup_iframe(config) {
        try {
            RED.nodes.createNode(this,config);
            var node = this;
            
            if (ui === undefined) {
                ui = RED.require("node-red-dashboard")(RED);
            }
            
            node.on('input', function(msg) {
                                
            });

            var done = ui.addWidget({
                node: node,
                format: generateHTML(node, config),
                templateScope: "local",
                group: config.group,
                emitOnlyNewValues: false,
                forwardInputMessages: false,
                storeFrontEndInputAsState: false,
                group: config.group,
                order: config.order,
                width: 0,
                height: -1,
                
                convertBack: function (value) {
                    return value;
                },

                beforeSend: function (msg, orig) {
                       
                },

                initController: function ($scope, events) {
                    let olddialog = document.querySelectorAll("#ui-popup-dialog_"+id);
                    olddialog.forEach(function(dialog) { dialog.remove(); } );
                    $scope.iframe = document.createElement("dialog");                   
                    setTimeout(function () {
                        //close if click outside
                        let onClick = function(event) {
                            if (event.target === $scope.iframe) {
                                $scope.dialogClose();
                            }
                        }
                        let dialog_style = document.querySelector("#ui-popup-dialog-style");
                        let iframe_style = document.querySelector("#ui-popup-iframe-style");
                        let iframe_style_id = document.querySelector("#ui-popup-iframe-style_"+id);
                        let cardpanel = document.querySelector("#ui-popup-iframe_" + id);
                        let cardPanelContainer = document.createElement("ui-card-panel");
                        document.head.appendChild(dialog_style);
                        document.head.appendChild(iframe_style);
                        document.head.appendChild(iframe_style_id);
                        cardpanel.parentElement.remove();
                        cardPanelContainer.appendChild(cardpanel);
                        cardPanelContainer.classList.add("ui-popup-iframe");
                        $scope.iframe.appendChild(cardPanelContainer);
                        $scope.iframe.setAttribute("id", "ui-popup-dialog_"+id);
                        $scope.iframe.classList.add("ui-popup-dialog");
                        $scope.iframe.classList.add("_"+id);
                        $scope.iframe.addEventListener("click", onClick);
                        document.body.appendChild($scope.iframe);
                    }, 300);
                    
                    $scope.dialogClose = function () {
                        $scope.iframe.close();
                        document.querySelector("#ui-popup-dialog_"+id+" iframe").removeAttribute("src");
                    }

                    $scope.$watch("msg", function (msg) {
                        if (!msg) { 
                            return;
                        } else {
                            if (msg.payload == "show") {
                                $scope.iframe.close();
                                $scope.iframe.showModal();
                                document.querySelector("#ui-popup-dialog_"+id+" iframe").setAttribute("src", document.querySelector("#ui-popup-dialog_"+id+" .content").getAttribute("framesrc"))
                            }
                        };                        
                    });
                },
                
                beforeEmit: function (msg, value) {
                    var newMsg = {};
		            if (msg) {
                        newMsg.socketid = msg.socketid;
                        newMsg.payload = msg.payload;
                    }
                    return { msg: newMsg };
                
            }

            });
        }
        catch (e) {
            //console.log(e);
        }
        node.on("close", done);
    }
    RED.nodes.registerType("ui_popup-iframe",ui_popup_iframe);
}